atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionManager.java

```diff
@@ -64,25 +64,21 @@ protected final <T, E extends Exception> T runTaskThrowOnConflict(TransactionTas
     protected final <T, E extends Exception> T runTaskThrowOnConflictWithCallback(
             TransactionTask<T, E> task, CallbackAwareTransaction txn, Runnable callback)
             throws E, TransactionFailedException {
+        checkOpen();
         try {
-            checkOpen();
-            try {
-                T ret = task.execute(txn);
-                if (txn.isUncommitted()) {
-                    txn.commitWithoutCallbacks();
-                }
-                return ret;
-            } finally {
-                // Make sure that anyone trying to retain a reference to this transaction
-                // will not be able to use it.
-                if (txn.isUncommitted()) {
-                    txn.abort();
-                }
+            txn.onCommitOrAbort(txn::reportExpectationsCollectedData);
+            txn.onCommitOrAbort(callback);
```

Comment:
Did we possibly change the ordering of these?

Reply:
Oh: we're inheriting the running order from the closer: shall we explicitly specify this in the javadoc that these callbacks will run in the reverse order to how they're registered?